aks.go

```diff
@@ -0,0 +1,173 @@
+)
+
+var (
+	ErrAzureTooManyTags error = errors.New("Only up to 50 tags can be set on an azure resource")
+	ErrAzureValueToLong error = errors.New("A value can only contain 256 characters")
+)
+
+type DiskTags = map[string]*string
+type AzureSubscription = string
+
+type AzureClient interface {
+	GetDiskTags(ctx context.Context, subscription AzureSubscription, resourceGroupName string, diskName string) (DiskTags, error)
+	SetDiskTags(ctx context.Context, subscription AzureSubscription, resourceGroupName string, diskName string, tags DiskTags) error
+}
+
+type azureClient struct {
+	client *armresources.TagsClient
+}
+
+func NewAzureClient() (AzureClient, error) {
+	creds, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return nil, err
+	}
+	client, err := armresources.NewTagsClient("", creds, &arm.ClientOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return azureClient{client}, err
+}
+
+func diskScope(subscription string, resourceGroupName string, diskName string) string {
+	return fmt.Sprintf("subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/disks/%s", subscription, resourceGroupName, diskName)
+}
+
+func (self azureClient) GetDiskTags(ctx context.Context, subscription AzureSubscription, resourceGroupName string, diskName string) (DiskTags, error) {
+
+	tags, err := self.client.GetAtScope(ctx, diskScope(subscription, resourceGroupName, diskName), &armresources.TagsClientGetAtScopeOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("could not get the tags for: %w", err)
+	}
+
+	return tags.Properties.Tags, nil
+}
+
+func (self azureClient) SetDiskTags(ctx context.Context, subscription AzureSubscription, resourceGroupName string, diskName string, tags DiskTags) error {
+	response, err := self.client.UpdateAtScope(
+		ctx,
+		diskScope(subscription, resourceGroupName, diskName),
+		armresources.TagsPatchResource{
+			to.Ptr(armresources.TagsPatchOperationReplace),
+			&armresources.Tags{Tags: tags},
+		}, &armresources.TagsClientUpdateAtScopeOptions{},
+	)
+	if err != nil {
+		return fmt.Errorf("could not set the tags for: %w", err)
+	}
+	log.WithFields(log.Fields{"disk": diskName, "resource-group": resourceGroupName}).Debugf("updated disk tags to tags=%v", response.Properties.Tags)
+	return nil
+}
+
+func parseAzureVolumeID(volumeID string) (subscription string, resourceGroup string, diskName string, err error) {
+	// '/subscriptions/{subscription}/resourceGroups/{resourceGroup}/providers/Microsoft.Compute/disks/{diskname}"'
+	fields := strings.Split(volumeID, "/")
+	if len(fields) != 9 {
+		return "", "", "", errors.New("invalid volume id")
+	}
+	subscription = fields[2]
+	resourceGroup = fields[4]
+	diskName = fields[8]
+	return subscription, resourceGroup, diskName, nil
+}
+
+func sanitizeLabelsForAzure(tags map[string]string) (DiskTags, error) {
+	diskTags := make(DiskTags)
+	if len(tags) > 50 {
+		return nil, ErrAzureTooManyTags
+	}
+	for k, v := range tags {
+		k = sanitizeKeyForAzure(k)
```

Comment:
After sanitizing the key it is possible that it could end up being a duplicate key. I'm on the fence as to whether it would be better to reject the tag if the key needs to be sanitized or if the tag created a duplicate. Which approach do you think is better?

Reply:
We use this in AWS and Azure so we don't really want to change the tags depending on the envrionment. 
But i will add an error in case there are duplicated keys, this should also tackle your other comment concerning `Kubernetes/Cluster: foo`